src/main/java/org/truffleruby/core/array/ArrayDupNode.java

```diff
@@ -29,7 +30,27 @@
 
     public abstract DynamicObject executeDup(VirtualFrame frame, DynamicObject array);
 
-    @Specialization(guards = "strategy.matches(from)", limit = "ARRAY_STRATEGIES")
+    @Specialization(guards = {
+            "strategy.matches(from)", "strategy.getSize(from) == cachedSize",
+            "cachedSize <= ARRAY_MAX_EXPLODE_SIZE"
+    }, limit = "3")
```

Comment:
Should we have an option for the limit?

Reply:
We can, but 8 is definitely too much here.

Reply:
+1 for option, better to be consistent with all other limits. (the value can be lower if needed)